File: lib/crdt/peer.rb
require 'openssl'

module CRDT
  # A pair of logical timestamp (Lamport clock, which is just a number) and peer ID (256-bit hex
  # string that uniquely identifies a particular device). A peer increments its timestamp on every
  # operation, so this pair uniquely identifies a particular object, e.g. an element in a list.
  # It also provides a total ordering that is consistent with causality: if operation A happened
  # before operation B, then A's ItemID is lower than B's ItemID. The ordering of concurrent
  # operations is deterministic but arbitrary.
  class ItemID < Struct.new(:logical_ts, :peer_id)
    include Comparable

    def <=>(other)
      return +1 if self.logical_ts > other.logical_ts
      return -1 if self.logical_ts < other.logical_ts
      self.peer_id <=> other.peer_id
    end
  end

  class Peer
    include Encoding

    # 256-bit hex string that uniquely identifies this peer.
    attr_reader :peer_id

    # Keeps track of the key facts that we know about our peers.
    attr_reader :peer_matrix

    # CRDT data structure (TODO generalise this)
    attr_reader :ordered_list

    # Lamport clock
    attr_reader :logical_ts

    # Loads a peer's state from a file with the specified +filename+ path.
    def self.load(filename)
      Encoding.load(filename)
    end

    # Initializes a new peer instance with default state. If no peer ID is given, it is assigned a
    # new random peer ID (256-bit hex string).
    def initialize(peer_id=nil)
      @peer_id = peer_id || bin_to_hex(OpenSSL::Random.random_bytes(32))
      @peer_matrix = PeerMatrix.new(@peer_id)
      @ordered_list = OrderedList.new(self)
      @logical_ts = 0
      @send_buf = []
      @recv_buf = {} # map of origin_peer_id => array of operations
    end

    def anything_to_send?
      !@peer_matrix.local_clock_update.empty? || !@send_buf.empty?
    end

    def next_id
      @logical_ts += 1
      ItemID.new(@logical_ts, peer_id)
    end

    def send_operation(operation)
      # Record causal dependencies of the operation before the operation itself
      if !peer_matrix.local_clock_update.empty?
        @send_buf << peer_matrix.local_clock_update
        peer_matrix.reset_clock_update
      end

      @send_buf << operation
      @peer_matrix.local_operation
    end

    # Returns a list of operations that should be sent to remote peers.
    # Resets the list, so the same operations won't be returned again.
    def flush_operations
      if !peer_matrix.local_clock_update.empty?
        @send_buf << peer_matrix.local_clock_update
        peer_matrix.reset_clock_update
      end

      return_ops = @send_buf
      @send_buf = []
      return_ops
    end

    # Receives operations from a remote peer. The operations will be applied immediately if they are
    # causally ready, or buffered until later if dependencies are missing.
    def receive_operations(origin_peer_id, operations)
      @recv_buf[origin_peer_id] ||= []
      @recv_buf[origin_peer_id].concat(operations)
      while apply_operations_if_ready; end
    end

    private

    # Checks if there are any causally ready operations in the receive buffer that we can apply, and
    # if so, applies them. Returns false if nothing was applied, and returns true if something was
    # applied. Keep calling this method in a loop until it returns false, to ensure all ready
    # buffers are drained.
    def apply_operations_if_ready
      ready_peer_id, ready_ops = @recv_buf.detect do |peer_id, ops|
        peer_matrix.causally_ready?(peer_id) && !ops.empty?
      end
      return false if ready_peer_id.nil?

      while ready_ops.size > 0
        operation = ready_ops.shift

        if operation.is_a? PeerMatrix::ClockUpdate
          peer_matrix.apply_clock_update(ready_peer_id, operation)

          # Applying the clock update might make the following operations causally non-ready, so we
          # stop processing operations from this peer and check again for causal readiness.
          return true
        else
          apply_operation(ready_peer_id, operation)
        end
      end

      true # Finished this peer, now another peer's operations might be causally ready
    end

    # Takes a single operation that we received from a remote peer, and applies it to the local
    # CRDT data structure.
    def apply_operation(origin_peer_id, operation)
      @logical_ts = operation.logical_ts if @logical_ts < operation.logical_ts
      peer_matrix.increment_op_count(origin_peer_id)
      ordered_list.apply_operation(operation)
    end
  end
end

File: java/src/main/java/org/trvedata/Peer.java
package org.trvedata;

import java.security.SecureRandom;
import java.util.ArrayDeque;
import java.util.Deque;
import java.util.HashMap;
import java.util.Map;

public class Peer<T> {
	private String peerId;
	private PeerMatrix peerMatrix;
	private OrderedList<T> orderedList;
	private long logicalTs;
	private Deque<Operation> sendBuf;
	private Map<String, Deque<Operation>> recvBuf;

	public Peer() {
		this(null);
	}

	public Peer(String peerId) {
		this.peerId = peerId != null ? peerId : this.createRandomPeerID();
		this.peerMatrix = new PeerMatrix(this.peerId);
		this.orderedList = new OrderedList<T>(this);
		this.logicalTs = 0;
		this.sendBuf = new ArrayDeque<Operation>();
		this.recvBuf = new HashMap<String, Deque<Operation>>();
	}

	protected String createRandomPeerID() {
		SecureRandom rand = new SecureRandom();
		byte[] bytes = new byte[32];
		String ret = "";
		rand.nextBytes(bytes);
		for (byte b : bytes)
			ret += String.format("%02x", b);
		return ret;
	}

	public boolean anythingToSend() {
		return !this.sendBuf.isEmpty();
	}

	protected ItemID nextId() {
		return new ItemID(++this.logicalTs, this.peerId);
	}

	protected void sendOperation(Operation operation) {
		this.sendClockUpdateIfNotEmpty();
		this.sendBuf.addLast(operation);
	}

	public Message makeMessage() {
		this.sendClockUpdateIfNotEmpty();
		Message message = new Message(this.peerId, this.peerMatrix.incrementMsgCount(), this.sendBuf);
		this.sendBuf = new ArrayDeque<Operation>();
		return message;
	}
	
	protected void sendClockUpdateIfNotEmpty() {
		if (!this.peerMatrix.getLocalClockUpdate().isEmpty()) {
			this.sendBuf.push(this.peerMatrix.getLocalClockUpdate());
			this.peerMatrix.resetClockUpdate();
		}
	}

	public void processMessage(Message message) {
		if (!(message instanceof Message))
			throw new RuntimeException("Invalid message: " + message);
		if (this.recvBuf.get(message.getOriginPeerId()) == null)
			this.recvBuf.put(message.getOriginPeerId(), new ArrayDeque<Operation>());
		// append all elements in message.operations to this.recvBuf
		this.recvBuf.get(message.getOriginPeerId()).addAll(message.getOperations());
		this.recvBuf.get(message.getOriginPeerId()).add(new MessageProcessed(message.getMsgCount()));
		while (this.applyOperationsIfReady())
			;
	}

	protected boolean applyOperationsIfReady() {
		String readyPeerId = null;
		Deque<Operation> readyOps = null;
		for (String peerId : this.recvBuf.keySet()) {
			if (this.peerMatrix.isCausallyReady(peerId) && !this.recvBuf.get(peerId).isEmpty()) {
				readyPeerId = peerId;
				readyOps = this.recvBuf.get(peerId);
				break;
			}
		}
		if (readyPeerId == null)
			return false;
		while (!readyOps.isEmpty()) {
			Operation operation = readyOps.pop();
			if (GlobalConstants.DEBUG)
				System.out.println("Peer " + peerId + ": Processing operation from " + readyPeerId + ": " + operation);
			if (operation instanceof ClockUpdate) {
				this.peerMatrix.applyClockUpdate(readyPeerId, (ClockUpdate) operation);
				// Applying the clock update might make the following operations causally non-ready, so we stop
				// processing operations from this peer and check again for causal readiness.
				return true;
			} else if (operation instanceof MessageProcessed) {
				MessageProcessed messageProcessed = (MessageProcessed) operation;
				this.peerMatrix.processedIncomingMsg(readyPeerId, messageProcessed.getMsgCount());
			} else {
				ChangingOperation changingOp = (ChangingOperation) operation;
				if (this.logicalTs < changingOp.logicalTs())
					this.logicalTs = changingOp.logicalTs();
				this.getOrderedList().applyOperation(changingOp);
			}
		}
		readyOps.clear();
		return true; // Finished this peer, now another peer's operations might be causally ready
	}

	public OrderedList<T> getOrderedList() {
		return orderedList;
	}

	public PeerMatrix getPeerMatrix() {
		return peerMatrix;
	}

	public String getPeerId() {
		return peerId;
	}

	@Override
	public String toString() {
		return "Peer [peerId=" + peerId + ", peerMatrix=" + peerMatrix + ", orderedList=" + orderedList + ", logicalTs=" + logicalTs + ", sendBuf="
				+ sendBuf + ", recvBuf=" + recvBuf + "]";
	}
}